Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp

```diff
@@ -0,0 +1,148 @@
+
+Mantid::HistogramData::HistogramY CreateMonteCarloWorkspace::fillHistogramWithRandomData(const std::vector<double> &cdf,
+                                                                                         int numIterations,
+                                                                                         std::mt19937 &gen) {
+  Mantid::HistogramData::HistogramY outputY(cdf.size(), 0.0);
+  std::uniform_real_distribution<> dis(0.0, 1.0);
+  for (int i = 0; i < numIterations; ++i) {
+    double randomNum = dis(gen);
+    // Find the bin corresponding to the random number in the CDF
+    auto it = std::lower_bound(cdf.begin(), cdf.end(), randomNum);
+    size_t index = std::distance(cdf.begin(), it);
+
+    // Ensure the index is within bounds
+    if (index < outputY.size()) {
+      outputY[index] += 1.0;
+    }
+  }
+  return outputY;
+}
+
+std::vector<double> CreateMonteCarloWorkspace::computeNormalizedCDF(const Mantid::HistogramData::HistogramY &yData) {
+  std::vector<double> cdf(yData.size());
+  std::partial_sum(yData.begin(), yData.end(), cdf.begin());
+  double total_counts = cdf.back();
+  // Normalize the CDF
+  std::transform(cdf.begin(), cdf.end(), cdf.begin(), [total_counts](double val) { return val / total_counts; });
+  return cdf;
+}
+
+int CreateMonteCarloWorkspace::computeNumberOfIterations(const Mantid::HistogramData::HistogramY &yData) {
+  double total_counts = std::accumulate(yData.begin(), yData.end(), 0.0);
+  return static_cast<int>(std::round(total_counts));
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+
+// Using Cumulative Distribution Function
+void CreateMonteCarloWorkspace::exec() {
+  // Progress Bar set up
+  int numSteps = 7;
+  API::Progress progress(this, 0.0, 1.0, numSteps);
+
+  MatrixWorkspace_sptr instWs = getProperty("InputWorkspace");
+  int seed_input = getProperty("Seed");
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  const Mantid::HistogramData::HistogramY &yData = instWs->y(0); // Counts in each bin
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  int numIterations = computeNumberOfIterations(yData);
+  std::vector<double> cdf = computeNormalizedCDF(yData);
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  MatrixWorkspace_sptr outputWs = WorkspaceFactory::Instance().create(instWs);
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  // Copy the bin boundaries (X-values) from the input to the output
+  outputWs->setSharedX(0, instWs->sharedX(0));
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  std::mt19937 gen(seed_input);
+
+  Mantid::HistogramData::HistogramY outputY = fillHistogramWithRandomData(cdf, numIterations, gen);
```

Comment:
I think it would be nicer if `fillHistogramsWithRandomData` took the seed as the input rather than a `std::mt19937`. There's no need to construct `gen` outside the method, since it's not sued after again after this point.